### PR TITLE
Fix TAL OOM recovery, AutoBatch loss estimation, LiteRT INT8 calibration, and nc validation

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Fix training OOM recovery, AutoBatch loss estimation, LiteRT INT8 calibration, and `nc` validation

Root-cause fixes for four failure classes surfaced by a combined Sentry sweep (both `alpha` and `ultralytics` projects, 24h window) and a 96h GPU training fleet log sweep on 2026-07-02. All failures reproduce on `8.4.84` (current `main`).

### Fixes

| Failure | Root cause | Fix | Type |
|---|---|---|---|
| Fatal `torch.OutOfMemoryError` inside the TaskAlignedAssigner CPU fallback (`tal.py:105` copy-back) — 17/17 failed fleet training jobs in the sweep window; Sentry ALPHA-1DM (343 events) | The CPU recompute and `.to(device)` copy-back ran **inside** the `except RuntimeError as e:` block, so `e.__traceback__` kept the failed `_forward` attempt's ~6 simultaneous `(bs, n_max_boxes, anchors)` fp32 GPU tensors alive during recovery — the copy-back then re-OOM'd allocating 14–34 GiB. The fallback never worked for exactly the workloads it exists for | Move recovery out of the `except` block; exiting it drops the traceback and returns the failed attempt's memory to the allocator before the CPU recompute and copy-back. Non-OOM `RuntimeError`s still re-raise | Replace |
| Residual OOM flavor: `batch=-1` picks batches the BCE classification loss can't afford (`loss.py:436`, 23.79 GiB alloc) on instance-dense / high-`nc` datasets | `profile_ops`' AutoBatch simulation allocated a single `(bs, max_num_obj, anchors)` tensor — the real detect loss holds ~6 simultaneous box-path buffers (TAL `get_box_metrics`) plus ~6 `(bs, anchors, nc)` cls-path fp32-equivalents | Simulate both peak envelopes (6× box + 6× cls) inside the same `cuda_memory_usage` block so the linear fit sees the true per-batch loss memory; errs toward smaller batches within the 0.6 fraction | Replace |
| Static-INT8 LiteRT export fails 100% deterministically with `No calibration batches of size N were collected` whenever the calibration dataset has fewer images than `args.batch` | `get_int8_calibration_dataloader` reduces the dataloader batch to `min(args.batch, n)` with `drop_last=True`, but the model was traced at `args.batch` — the size-match skip in `torch2litert` then collected zero samples | Tile under-sized batches up to the traced batch dim (repeats are statistics-identical for calibration). The assert became unreachable dead code and was deleted | Replace + Delete |
| `check_det_dataset` rejects valid YAML with the self-contradictory message `'names' length 10 and 'nc: 10' must match` (7% of ULTRALYTICS-2151 samples); names-absent quoted `nc` crashes at `range(str)` | `len(data["names"]) != data["nc"]` does no type coercion, so a YAML-quoted `nc: "10"` (str) never equals int 10 | Coerce `nc` to int once before the guard, accepting integer-like values (`10`, `"10"`, `10.0`) and rejecting non-integral/placeholder values (`1.9`, `number_of_classes`) with a clear `'nc: ...' must be an integer` error | Add (5 lines) |

Also bumps `__version__` to **8.4.85** for release.

### Verification

- `tests/test_python.py -k profile` + a direct `profile_ops(max_num_obj=50)` smoke run exercising the new simulation branch
- Live `check_det_dataset` probes: accepts `10` / `"10"` / `10.0`, rejects `1.9` / `10.0`-vs-9-names mismatch / `number_of_classes` / non-numeric; targeted pytest (7 passed)
- `ruff check` + `ruff format` clean on all changed files

### Review

Independent dual review, both LGTM: Claude Code reviewer (4 rounds, adversarial) and Codex CLI (`gpt-5.5`, high reasoning, 3 rounds — earlier rounds drove the 6× envelope scaling and the integral-float `nc` semantics). Sole remaining note (`nc: .inf` → `OverflowError`) judged non-blocking pathological-input polish by both reviewers.

Cross-reference: companion platform-side PR (worker harness fixes from the same sweep): https://github.com/ultralytics/portal/pull/2656

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 This PR bumps Ultralytics to `8.4.85` and improves dataset validation, LiteRT export calibration, CUDA OOM recovery, and AutoBatch memory profiling.

### 📊 Key Changes
- 📦 Updated package version from `8.4.84` to `8.4.85`.
- ✅ Strengthened dataset YAML validation by ensuring `nc` is an integer, while still accepting integer-like values such as `"10"` or `10.0`.
- 📱 Improved LiteRT quantization calibration by repeating smaller calibration batches instead of failing when the final batch is undersized.
- 🧠 Refined CUDA out-of-memory fallback in `TaskAlignedAssigner` to release failed GPU traceback memory before retrying on CPU.
- 📊 Enhanced AutoBatch profiling memory simulation to better reflect detection-loss memory peaks during training.
- 📝 Minor documentation and test docstring formatting updates for LiteRT benchmarks and CUDA device selection.

### 🎯 Purpose & Impact
- 🛡️ Helps users catch invalid dataset configs earlier with clearer errors, reducing confusing training failures.
- 📲 Makes LiteRT export more robust, especially for small datasets or partial calibration batches.
- 💪 Improves training stability when GPU memory is tight by making CPU fallback more reliable after CUDA OOM errors.
- ⚖️ Provides more realistic memory estimates for AutoBatch, helping users choose safer batch sizes.
- ✨ Includes small polish updates to docs and tests for consistency and readability.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message is here to help with review readiness, and an Ultralytics engineer will assist as needed. Your PR appears to improve TAL OOM recovery, AutoBatch memory estimation, LiteRT INT8 calibration, and dataset `nc` validation 🔧

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/export/litert.py [hunk 1]
```diff
@@ -121,13 +121,11 @@ def torch2litert(
...
-                # litert-torch traces a fixed batch; feed whole batches matching im's batch dim (skip a partial tail).
-                if imgs.shape[0] == im.shape[0]:
-                    calib_samples.append({"args_0": imgs.numpy()})
-            assert calib_samples, (
-                f"No calibration batches of size {im.shape[0]} were collected; the calibration dataset needs at least "
-                f"'batch={im.shape[0]}' images for static LiteRT quantization (reduce 'batch' or add more 'data')."
-            )
+                # litert-torch traces a fixed batch; tile under-sized batches up to im's batch dim (repeats are
+                # statistics-identical for calibration)
+                if imgs.shape[0] < im.shape[0]:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/litert.py:126]: 💡 **MEDIUM**: If the calibration dataloader yields an empty batch, this branch divides by `imgs.shape[0]` and then attempts `repeat(... // 0)`, which will fail with a low-level `ZeroDivisionError`. The previous `assert calib_samples` produced a clearer failure when no valid samples were collected; this path should skip or explicitly reject empty batches before tiling.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR is generally well-scoped and the TAL OOM recovery plus AutoBatch memory simulation changes look reasonable from the diff. One edge case in LiteRT calibration needs attention: empty calibration batches now cause a division-by-zero during tiling instead of a clear validation failure.

💬 Posted 1 inline comment

File: ultralytics/utils/export/litert.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/utils/export/litert.py:126]: Empty batches are impossible on this path: `get_int8_calibration_dataloader` raises `ValueError` for datasets with n<1 images and builds the loader with `batch=min(args.batch, n)>=1` and `drop_last=True`, so every yielded batch has `imgs.shape[0]>=1` — the divisor can never be 0. The deleted assert guarded the old size-mismatch skip (which could collect zero samples); after tiling, every batch is appended, so both the empty-batch and empty-samples states are unreachable. Adding a guard here would be error handling for an impossible state.

**Comment by UltralyticsAssistant** [comment]
🚀 Merged — thank you @glenn-jocher!

“Success is the sum of small efforts, repeated day in and day out.” — Robert Collier

This release bump to `8.4.85` is a great example of that principle in action: stronger dataset validation, more resilient LiteRT calibration, improved CUDA OOM recovery, and more realistic AutoBatch memory profiling all add up to a smoother and more reliable Ultralytics experience. These refinements help users catch issues earlier, train more confidently under memory pressure, and export models more robustly across real-world datasets and devices.

Appreciate the continued polish and care that keeps Ultralytics moving forward for the community 🙌
